File: app/lexer/token.py
"""Token Data Structures"""

from enum import Enum, auto
from dataclasses import dataclass

class TokenType(Enum):
    PAREN_OPEN = auto()
    PAREN_CLOSE = auto()
    CURLY_OPEN = auto()
    CURLY_CLOSE = auto()
    SQUARE_OPEN = auto()
    SQUARE_CLOSE = auto()
    COLON = auto()
    COMMA = auto()
    KEYWORD = auto()
    INTEGER = auto()
    FLOAT = auto()
    STRING = auto()
    OPERATOR = auto()
    IDENTIFIER = auto()
    COMMENT = auto()
    RESERVED = auto()

@dataclass
class Token:
    type: TokenType
    value: str
    start_line: int = 0
    end_line: int = 0
    start_column: int = 0
    end_column: int = 0

    def __repr__(self):
        return "Token({}, '{}', line: {}-{}, column: {}-{})".format(self.type.name, self.value, self.start_line, self.end_line, self.start_column, self.end_column)

@dataclass
class TokenizeResult:
    consumed_chars: int
    token: Token

@dataclass
class LexContext:
    data: list[str]
    index: int

    @property
    def current(self):
        return self.data[self.index]

    def char_at_offset(self, offset: int):
        return self.data[self.index + offset]

File: app_nim/lexer/token.nim
import std/options
import std/strformat

type
  TokenType* = enum
    PAREN_OPEN,
    PAREN_CLOSE,
    CURLY_OPEN,
    CURLY_CLOSE,
    SQUARE_OPEN,
    SQUARE_CLOSE,
    COLON,
    COMMA,
    KEYWORD,
    INTEGER,
    FLOAT,
    STRING,
    OPERATOR,
    IDENTIFIER,
    COMMENT,
    RESERVED

  Token* = ref object
    tok_type*: TokenType
    value*: string
    start_line*: int = 0
    end_line*: int = 0
    start_column*: int = 0
    end_column*: int = 0

  TokenizeResult* = object
    consumed_chars*: int
    token*: Option[Token]

  # context type for easily passing data around
  LexContext* = ref object
    data*: seq[char]
    idx*: int

proc has_next*(ctx: LexContext): bool =
  result = ctx.idx < len(ctx.data)
proc consume*(ctx: LexContext, n: int): void =
  ctx.idx = ctx.idx + n

proc current*(ctx: LexContext): char =
  return ctx.data[ctx.idx]

proc peek*(ctx: LexContext, offset: int): char =
  return ctx.data[ctx.idx + offset]

proc `$`*(t: Token): string =
  result = &"Token(type: {t.tok_type}, value: '{t.value}', lines: {t.start_line}-{t.end_line}, cols: {t.start_column}-{t.end_column})"